### PR TITLE
Allow assigning ArrayInterval to ArrayInterval and improvements to Exceptions of ArrayInterval

### PR DESCRIPTION
**Pull request by thequilo**

### CODE REVIEW

File: paderbox/array/interval/core.py [hunk 1]
```diff
@@ -466,33 +474,66 @@ def __setitem__(self, item, value):
...
-        elif isinstance(value, (tuple, list, np.ndarray)):
-            assert len(value) == stop - start, (start, stop, stop - start, len(value), value)
-            ai = ArrayInterval(value, inverse_mode=self.inverse_mode)
-            intervals = self.intervals
-            intervals = cy_non_intersection((start, stop), intervals)
-            self.intervals = intervals + tuple([(s + start, e + start) for s, e in ai.intervals])
+                self.intervals = cy_non_intersection((start, stop), self.intervals)
+        elif isinstance(value, (tuple, list, np.ndarray, ArrayInterval)):
+            if not isinstance(value, ArrayInterval):
+                value = ArrayInterval(value, inverse_mode=self.inverse_mode)
```
Comment (boeddeker) [inline paderbox/array/interval/core.py:517]: Why `inverse_mode=self.inverse_mode` and not `inverse_mode=False`?

Comment (boeddeker) [inline paderbox/array/interval/core.py:517]: Can you add a comment, why `inverse_mode=self.inverse_mode`?

File: paderbox/array/interval/core.py [hunk 2]
```diff
@@ -466,33 +474,66 @@ def __setitem__(self, item, value):
...
+
+            if len(value) != stop - start:
+                raise ValueError(
+                    f'Could not broadcast input with length {len(value)} into '
+                    f'shape {stop - start}'
+                )
+            if value.inverse_mode != self.inverse_mode:
+                raise ValueError(
```
Comment (boeddeker) [inline paderbox/array/interval/core.py:525]: Maybe a `NotImplementedError`?

File: paderbox/array/interval/core.py [hunk 3]
```diff
@@ -514,16 +555,33 @@ def __getitem__(self, item):
...
+        >>> ai[-25:-20]
+        array([ True,  True,  True,  True,  True])
+        >>> ai[-1:]
+        array([])
+
         """
         if isinstance(item, (int, np.integer)):
+            index = item
+            if index < 0:
+                index = index + self.shape[0]
```
Comment (boeddeker) [inline paderbox/array/interval/core.py:567]: `.shape[0]` -> `.shape[-1]` Just for the case, we support independent dims in the furture

Comment (boeddeker) [reply paderbox/array/interval/core.py:567]: Shape may be None. Can you catch this case?

File: paderbox/array/interval/core.py [hunk 4]
```diff
@@ -466,33 +474,83 @@ def __setitem__(self, item, value):
...
+                f'{self.__class__.__name__}.__setitem__ only supports slices '
+                f'for indexing, not {item}'
+            )
 
         start, stop = cy_parse_item(item, self.shape)
 
         if np.isscalar(value):
+            if value not in (0, 1):
+                # Numpy interprets values as boolean even if they are larger
+                # than 1, why doesn't the ArrayInterval do the same?
```
Comment (boeddeker) [inline paderbox/array/interval/core.py:519]: Checking if the value is a Boolean (or 0, 1) is more verbose, and I don't know a case, where other values are useful. Usually they indicate a bug or a wrong assumption of the user.

File: paderbox/array/interval/core.py (hunk 1 above)
Comment (thequilo) [reply paderbox/array/interval/core.py:533]: Done

File: paderbox/array/interval/core.py (hunk 2 above)
Comment (thequilo) [reply paderbox/array/interval/core.py]: Done

File: paderbox/array/interval/core.py (hunk 3 above)
Comment (thequilo) [reply paderbox/array/interval/core.py]: Done

**Review by boeddeker — APPROVED** [review]

File: paderbox/array/interval/core.py (hunk 4 above)
Comment (thequilo) [reply paderbox/array/interval/core.py]: I updated the comment

**Pull request merged**